firefox_puppeteer/api/software_update.py

```diff
@@ -0,0 +1,544 @@
+        """
+
+        url = self.prefs.get_pref(self.PREF_APP_UPDATE_URL, '')
+        # get the next two prefs from the default branch
+        dist = self.prefs.get_pref(self.PREF_APP_DISTRIBUTION, True) or 'default'
+        dist_version = self.prefs.get_pref(self.PREF_APP_DISTRIBUTION_VERSION, True) or 'default'
+
+        if not url:
+            return None
+
+        # Not all placeholders are getting replaced correctly by formatURL
+        url = re.sub(r'%PRODUCT%', self.app_info.name, url)
+        url = re.sub(r'%BUILD_ID%', self.app_info.appBuildID, url)
+        url = re.sub(r'%BUILD_TARGET%', self.app_info.OS + '_' + self.ABI, url)
+        url = re.sub(r'%OS_VERSION%', self.os_version, url)
+        url = re.sub(r'%CHANNEL%', self.update_channel.channel, url)
+        url = re.sub(r'%DISTRIBUTION%', dist, url)
+        url = re.sub(r'%DISTRIBUTION_VERSION%', dist_version, url)
+
+        url = self.marionette.execute_script("""
+            Cu.import("resource://gre/modules/Services.jsm");
+
+            var url = arguments[0];
+            return Services.urlFormatter.formatURL(url);
+
+            """, script_args=[url])
+
+        if force:
+            if '?' in url:
+                url += '&'
+            else:
+                url += '?'
+            url += 'force=1'
+        return url
+
+
+class ActiveUpdate(object):
+
+    def __init__(self, software_update):
+
+        self.marionette = software_update.marionette
+
+    def __getattr__(self, attr):
+        value = self.marionette.execute_script("""
+            let ums = Cc['@mozilla.org/updates/update-manager;1']
+                .getService(Ci.nsIUpdateManager);
+            var active_update = ums.activeUpdate;
+            return active_update[arguments[0]];
+        """, script_args=[attr])
+
+        if value is not None:
+            return value
+        else:
+            raise AttributeError('{} has no attribute {}'.format(self.__class__.__name__,
+                                                                 attr))
+
+    @property
+    def exists(self):
+        """Is there an active update
+
+        :returns: True if there is an active update
+        """
+
+        active_update = self.marionette.execute_script("""
+            let ums = Cc['@mozilla.org/updates/update-manager;1']
+                .getService(Ci.nsIUpdateManager);
+            return ums.activeUpdate;
+
+        """)
+
+        return bool(active_update)
+
+    def get_patch_at(self, patch_index):
+        """Use nsIUpdate.getPatchAt to return a patch from an update
+
+        :returns: nsIUpdatePatch object
+        """
+
+        return self.marionette.execute_script("""
+            let ums = Cc['@mozilla.org/updates/update-manager;1']
+                .getService(Ci.nsIUpdateManager);
+            let active_update = ums.activeUpdate;
+            let patch_index = arguments[0];
+            return active_update.getPatchAt(patch_index);
```

Comment:
I checked DXR and so far we really only use properties so we should be fine in returning the JSON data for the interface.